app/src/org/commcare/android/tasks/TemplatePrinterTask.java

```diff
@@ -0,0 +1,239 @@
+
+    /**
+     * The 2 types of errors that can be encountered in this task
+     */
+    public static final int IO_ERROR = 0;
+    private static final int VALIDATION_ERROR = 1;
+
+    /**
+     * The template file for this print action
+     */
+    private final File inputFile;
+
+    /**
+     * The path where the populated template should be saved to
+     */
+    private final String outputPath;
+
+    /**
+     * The mapping from keywords to case property values to be used in populating the template
+     */
+    private final Bundle values;
+
+    private final PopulateListener listener;
+
+
+    public TemplatePrinterTask(File input, String outputPath, Bundle values,
+                               PopulateListener listener) {
+        this.inputFile = input;
+        this.outputPath = outputPath;
+        this.values = values;
+        this.listener = listener;
+    }
+
+    /**
+     * Attempts to perform population of the template file, and throws the appropriate exception
+     * if encountering an error.
+     */
+    @Override
+    protected Boolean doInBackground(Void... params) {
+        try {
+            populateHtml(inputFile, values);
+            return true;
+        } catch (IOException e) {
+            errorType = IO_ERROR;
+            return false;
+        } catch (RuntimeException e) {
+            errorType = VALIDATION_ERROR;
+            return false;
+        }
+    }
+
+    /**
+     * Receives the return value from doInBackground and proceeds accordingly
+     */
+    @Override
+    protected void onPostExecute(Boolean success) {
+        if (success) {
+            listener.onFinished();
+        } else {
+            listener.onError(errorType);
+        }
+    }
+
+    /**
+     * Populates an html print template based on the given set of key-value pairings
+     * and save the newly-populated template to a temp location
+     *
+     * @param input the html print template
+     * @param values the mapping of keywords to case property values
+     * @throws IOException
+     */
+    private void populateHtml(File input, Bundle values) throws IOException {
+        // Read from input file
+        // throws IOException
+        String fileText = TemplatePrinterUtils.docToString(input);
+
+        // Check if <body></body> section of html string is properly formed
+        // throws RuntimeException
+        int startBodyIndex = fileText.indexOf("<body");
+        String beforeBodySection = fileText.substring(0, startBodyIndex);
+        String bodySection = fileText.substring(startBodyIndex);
+        validateStringOrThrowException(bodySection);
+
+        // Swap out place-holder keywords for case property values within <body></body> section
+        bodySection = replace(bodySection, values);
+
+        // Write the new HTML to the desired  temp file location
+        // throws IO Exception
+        TemplatePrinterUtils.writeStringToFile(beforeBodySection + bodySection, outputPath);
```

Comment:
The <body></body> section of the html doc is the only place where a user can put any content, so it's only necessary to call replace() on that section of the document. Therefore calling either validateString() or replace() on the entire file string is a waste of time (the validate method is pretty slow since it goes character by character), and also unnecessarily introduces the possibility of having a validation error thrown when we don't want it. Even if there was something that we consider "illegal" in the <head> section, we don't need to care about it, so it doesn't make sense to allow that to happen.